torch/csrc/jit/codegen/cuda/ir_nodes.cpp

```diff
@@ -1031,6 +1030,325 @@ std::pair<TensorDomain*, TensorDomain*> TensorDomain::rFactor(
+  //!          will create a new equivalent class if b does
+  //!          not belong to any
+  void join(T a, T b) {
+    const int i0 = fixedPoint(a);
+    const int i1 = fixedPoint(b);
+    int new_parent, new_child;
+
+    // either order here is correct but joining larger class to smaller class
+    // tend to be faster
+    std::tie(new_parent, new_child) = (weights[i0] < weights[i1])
+        ? std::make_pair(i0, i1)
+        : std::make_pair(i1, i0);
+    weights[new_parent] += weights[new_child];
+    set_map[new_child] = new_parent;
+  }
+
+  //! Checks if a and b belong to the same equivalent class
+  //!
+  //! \param a an element from a equivalent class
+  //! \param b an element from another equivalent class
+  //! \returns boolean value representing if a and b are
+  //!          recorded to be in the same equivalent class
+  //!          will return false if any of a or b doesn't
+  //!          have an equivalent class recorded
+  bool areEqual(T a, T b) {
+    if (!entry_map.count(a) || !entry_map.count(b))
+      return false;
+    return fixedPoint(a) == fixedPoint(b);
+  }
+
+ private:
+  // Internal representation of the equivalence class as integers
+  // set_map implements the "parent" relationship
+  std::vector<int> set_map;
+  // weights is used for preliminary perf optimization
+  std::vector<int> weights;
+
+  // Map the input of type T to its equivalence class
+  std::unordered_map<T, int> entry_map;
+
+  // Utility for generating new class
+  int count = 0;
+
+  // internal fixed point implementation:
+  //  returns the equivalent class that e belongs to
+  //  does adhoc optimization to speed up future access
+  int fixedPoint(int e) {
+    TORCH_INTERNAL_ASSERT(set_map.size() > e);
+    while (set_map[e] != e) {
+      // optimization step
+      set_map[e] = set_map[set_map[e]];
+      // chasing to fixed point
+      e = set_map[e];
+    }
+    return e;
+  }
+
+  //! utility to check the class i belongs to:
+  //!
+  //! will create a new class if no match seen
+  //! \param i element i to find the equiv class for
+  //! \returns the equivalent class that e belongs to
+  //!
+  int fixedPoint(T i) {
+    // handles case when i doesn't have an equivalence class
+    if (!entry_map.count(i))
+      createPoint(i);
+    // use fixed point as a representation for the equiv class
+    return fixedPoint(entry_map[i]);
+  }
+
+  //! utility to create a new equiv class for i
+  //
+  //! \param i element i to create the equiv class for
+  void createPoint(T i) {
+    entry_map[i] = count;
+    set_map.push_back(count++);
+    weights.push_back(1);
+  }
+};
+
+//! Concretize broadcast axes, i.e. identifying a non-broadcast
+//! IterDomain that the broadcast IterDomain can map to.
+//!
+//! This traversal processes root domains only, concretization works by
+//! inspecting pointwise ops, e.g. : T2 [i0,i1] = T1[i0,B0] + T0[i0,i1]
+//! will concretize axis B0 to i1
+//!
+class ConcretizeDomain : public BackwardVisitor {
```

Comment:
we can use private inheritance here (`private BackwardVisitor`) since BackwardVisitor is just a mixing (ie. we never want to tread `ConcretizeDomain` as a `BackwardVisitor`, right?)

Reply:
updated. Thanks!